core/multigrid/amgx_pgm.cpp

```diff
@@ -0,0 +1,193 @@
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+
+
+#include "core/components/fill_array.hpp"
+#include "core/matrix/csr_builder.hpp"
+#include "core/multigrid/amgx_pgm_kernels.hpp"
+
+
+namespace gko {
+namespace multigrid {
+namespace amgx_pgm {
+
+
+GKO_REGISTER_OPERATION(restrict_apply, amgx_pgm::restrict_apply);
+GKO_REGISTER_OPERATION(prolong_applyadd, amgx_pgm::prolong_applyadd);
+GKO_REGISTER_OPERATION(match_edge, amgx_pgm::match_edge);
+GKO_REGISTER_OPERATION(count_unagg, amgx_pgm::count_unagg);
+GKO_REGISTER_OPERATION(renumber, amgx_pgm::renumber);
+GKO_REGISTER_OPERATION(find_strongest_neighbor,
+                       amgx_pgm::find_strongest_neighbor);
+GKO_REGISTER_OPERATION(assign_to_exist_agg, amgx_pgm::assign_to_exist_agg);
+GKO_REGISTER_OPERATION(amgx_pgm_generate, amgx_pgm::amgx_pgm_generate);
+GKO_REGISTER_OPERATION(fill_array, components::fill_array);
+
+
+}  // namespace amgx_pgm
+
+
+namespace {
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> amgx_pgm_generate(
+    std::shared_ptr<const Executor> exec,
+    const matrix::Csr<ValueType, IndexType> *source, const size_type num_agg,
+    const Array<IndexType> &agg)
+{
+    auto coarse = matrix::Csr<ValueType, IndexType>::create(
+        exec, dim<2>{num_agg, num_agg}, 0, source->get_strategy());
+    exec->run(amgx_pgm::make_amgx_pgm_generate(source, agg, coarse.get()));
+    return std::move(coarse);
+}
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void AmgxPgm<ValueType, IndexType>::generate()
+{
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+    using rmc_value_type = remove_complex<ValueType>;
+    using weight_matrix_type = remove_complex<matrix_type>;
+    auto exec = this->get_executor();
+    const auto num = this->system_matrix_->get_size()[0];
+    Array<IndexType> strongest_neighbor(this->get_executor(), num);
+    Array<IndexType> intermediate_agg(this->get_executor(),
+                                      parameters_.deterministic * num);
+    // Only support csr matrix currently.
+    const matrix_type *amgxpgm_op = nullptr;
+    // Store the csr matrix if needed
+    auto amgxpgm_op_unique_ptr = matrix_type::create(exec);
+    if (!(amgxpgm_op =
+              dynamic_cast<const matrix_type *>(system_matrix_.get()))) {
+        // if original matrix is not csr, converting it to csr.
+        as<ConvertibleTo<matrix_type>>(this->system_matrix_.get())
+            ->convert_to(amgxpgm_op_unique_ptr.get());
+        amgxpgm_op = amgxpgm_op_unique_ptr.get();
+    }
+
+    // Initial agg = -1
+    exec->run(amgx_pgm::make_fill_array(agg_.get_data(), agg_.get_num_elems(),
+                                        -one<IndexType>()));
+    size_type num_unagg{0};
+    size_type num_unagg_prev{0};
+    // TODO: if mtx is a hermitian matrix, weight_mtx = abs(mtx)
```

Comment:
Is this TODO meant for later, to avoid the transposition if we know it is symmetric/hermitian?

Reply:
Yes, it will need the generic Linear Operator properties.
another way is to do a similar thing `is_hermitian and is_symmetric` like `skip_sorting` 
- is_symmetric: (A + A^T)/2 = A
- is_hermitian: (A + A^T)/2 = real(A)
- (Maybe) consider directed weight as non-directed weight: not need (A+A^T)/2. Not sure it makes sense or not.